libkernelflinger/security.c

```diff
@@ -216,6 +233,107 @@ struct rot_data_t* get_rot_data()
 	return &rot_data;
 }
 
+/* Update the struct attestation_ids for startup_information */
+EFI_STATUS update_attestation_ids(IN VOID *vendorbootimage)
+{
+    EFI_STATUS ret = EFI_SUCCESS;
+
+    struct vendor_boot_img_hdr_v4 *vendor_hdr = (struct vendor_boot_img_hdr_v4 *)vendorbootimage;
+    UINT32 page_size = vendor_hdr->page_size;
+    UINT32 bootconfig_offset = ALIGN(sizeof(struct vendor_boot_img_hdr_v4), page_size) +
+                               ALIGN(vendor_hdr->vendor_ramdisk_size, page_size) +
+                               ALIGN(vendor_hdr->dtb_size, page_size) +
+                               ALIGN(vendor_hdr->vendor_ramdisk_table_size, page_size);
+    /* Initialize the attestation ids structure */
+    if (vendor_hdr->bootconfig_size) {
+        UINT8 *configChar = AllocatePool(vendor_hdr->bootconfig_size + 1);
+        memcpy_s(configChar,
+        vendor_hdr->bootconfig_size, vendorbootimage + bootconfig_offset,
+        vendor_hdr->bootconfig_size);
+        configChar[vendor_hdr->bootconfig_size] = '\0';
+        debug(L"### bootconfig %a", configChar);
+        const char *delim = "\n";
+        CHAR8 *savedPtr;
+        CHAR8 *token = (CHAR8 *)strtok_r((char *)configChar, delim, (char **)&savedPtr);
+        const char *delim2 = "=";
+        CHAR8 *savedPtr2;
+        const char *d1 = ",";
+        const char *d2 = " ";
+        while (token != NULL) {
+            if (strncmp(token, "androidboot.brand", strlen("androidboot.brand")) == 0) {
+                CHAR8 *temp_brand = strtok_r(token, delim2, (char **)&savedPtr2);
+                temp_brand = strtok_r(NULL, delim2, (char **)&savedPtr2);//Get the string after the equal sign
+                temp_brand = strrpl(temp_brand, d1, d2);//Converts comma to space in the string
+                attestation_ids.brandSize = strlen(temp_brand);
+                CopyMem(attestation_ids.brand, temp_brand, attestation_ids.brandSize);
+            } else if(strncmp(token, "androidboot.device", strlen("androidboot.device")) == 0) {
```

Comment:
suggest to slipt the line #264~line#268 into a function(e.g. set_attestation_ids(char8*, char8*, uint32_t)), they're duplicated in many places